donkeycar/parts/pins.py

```diff
@@ -0,0 +1,913 @@
+        pin_provider = parts[0]
+        if PinScheme.BCM != parts[1]:
+            raise ValueError("Pin scheme must be BCM for PIGPIO")
+        pin_number = int(parts[2])
+        return input_pin(pin_provider, pin_number, pin_scheme=PinScheme.BCM, pull=pull)
+
+    raise ValueError("Unknown pin provider {}".format(parts[0]))
+
+
+def input_pin(pin_provider:str, pin_number:int, pin_scheme:str=PinScheme.BOARD, pull:int=PinPull.PULL_NONE) -> InputPin:
+    """
+    construct an InputPin using the given pin provider
+    """
+    if pin_provider == PinProvider.RPI_GPIO:
+        return InputPinGpio(pin_number, pin_scheme, pull)
+    if pin_provider == PinProvider.PCA9685:
+        raise RuntimeError("PinProvider.PCA9685 does not implement InputPin")
+    if pin_provider == PinProvider.PIGPIO:
+        if pin_scheme != PinScheme.BCM:
+            raise ValueError("Pin scheme must be PinScheme.BCM for PIGPIO")
+        return InputPinPigpio(pin_number, pull)
+    raise RuntimeError("UnknownPinProvider ({})".format(pin_provider))
+
+
+def output_pin(pin_provider:str, pin_number:int, pin_scheme:str=PinScheme.BOARD, i2c_bus:int=0, i2c_address:int=40, frequency_hz:int=60) -> OutputPin:
+    """
+    construct an output pin using the given pin provider
+    """
+    if pin_provider == PinProvider.RPI_GPIO:
+        return OutputPinGpio(pin_number, pin_scheme)
+    if pin_provider == PinProvider.PCA9685:
+        return OutputPinPCA9685(pin_number, pca9685(i2c_bus, i2c_address, frequency_hz))
+    if pin_provider == PinProvider.PIGPIO:
+        if pin_scheme != PinScheme.BCM:
+            raise ValueError("Pin scheme must be PinScheme.BCM for PIGPIO")
+        return OutputPinPigpio(pin_number)
+    raise RuntimeError("UnknownPinProvider ({})".format(pin_provider))
+
+
+def pwm_pin(pin_provider:str, pin_number:int, pin_scheme:str=PinScheme.BOARD, frequency_hz:int=60, i2c_bus:int=0, i2c_address:int=40) -> PwmPin:
+    """
+    construct a PwmPin using the given pin provider
+    """
+    if pin_provider == PinProvider.RPI_GPIO:
+        return PwmPinGpio(pin_number, pin_scheme, frequency_hz)
+    if pin_provider == PinProvider.PCA9685:
+        return PwmPinPCA9685(pin_number, pca9685(i2c_bus, i2c_address, frequency_hz))
+    if pin_provider == PinProvider.PIGPIO:
+        if pin_scheme != PinScheme.BCM:
+            raise ValueError("Pin scheme must be PinScheme.BCM for PIGPIO")
+        return PwmPinPigpio(pin_number, frequency_hz)
+    raise RuntimeError("UnknownPinProvider ({})".format(pin_provider))
+
+
+#
+# ----- RPi.GPIO/Jetson.GPIO implementations -----
+#
+try:
+    import RPi.GPIO as GPIO
+    # lookups to convert abstact api to GPIO values
+    gpio_pin_edge = [None, GPIO.RISING, GPIO.FALLING, GPIO.BOTH]
+    gpio_pin_pull = [None, GPIO.PUD_OFF, GPIO.PUD_DOWN, GPIO.PUD_UP]
+    gpio_pin_scheme = {PinScheme.BOARD: GPIO.BOARD, PinScheme.BCM: GPIO.BCM}
+except ImportError:
+    print("pigpio was not imported.")
```

Comment:
We switched to logging for these kind of shell outputs.

Reply:
I've changed print() to logger.debug()

Reply:
not here, but I saw your changes in the other files.

Reply:
Fixed in pins.py now.  Thanks.